File: ecosystem/sdk-php/src/Revenue.php
<?php
//
// Revenue.php
// The money one event reports.
//
// Created: 2026-08-30
// Copyright (c) 2026 Cloudmanic Labs, LLC. All rights reserved.
//

declare(strict_types=1);

namespace Feasible;

use Feasible\Exception\InvalidEventException;

/**
 * The `$` field of the wire payload. It is a type rather than a loose array so
 * that a currency typo fails at the call site: the server quietly ignores a
 * revenue object with no currency, and a revenue report that is silently zero
 * is the hardest kind of missing data to notice.
 */
final class Revenue
{
    /** The amount in major units, the way a payment provider reports a total. */
    public readonly int|float $amount;

    /** The ISO 4217 code, upper-cased. */
    public readonly string $currency;

    /**
     * Normalises the currency to the upper-case form the server stores, so that
     * "usd" and "USD" do not become two rows on the same report.
     */
    public function __construct(int|float $amount, string $currency)
    {
        $code = strtoupper(trim($currency));

        if (preg_match('/^[A-Z]{3}$/', $code) !== 1) {
            throw new InvalidEventException(sprintf(
                'revenue currency %s is not a three-letter ISO 4217 code, such as USD or GBP',
                var_export($currency, true)
            ));
        }

        $this->amount = $amount;
        $this->currency = $code;
    }

    /**
     * Renders the wire shape. The key names are the server's and are not
     * configurable, so they live in exactly one place.
     *
     * @return array{amount: int|float, currency: string}
     */
    public function toArray(): array
    {
        return ['amount' => $this->amount, 'currency' => $this->currency];
    }
}

File: ecosystem/sdk-ruby/lib/feasible/revenue.rb
#
# revenue.rb
# The money one event reports.
#
# Created: 2026-08-30
# Copyright (c) 2026 Cloudmanic Labs, LLC. All rights reserved.
#

# frozen_string_literal: true

module Feasible
  # The `$` field of the wire payload. It is a class rather than a loose hash so
  # a currency typo fails at the call site: the server ignores a revenue object
  # with no currency, and revenue that is silently zero is the hardest kind of
  # missing data to notice.
  class Revenue
    attr_reader :amount, :currency

    # Normalises the currency to the upper-case ISO 4217 form the server stores,
    # so that "usd" and "USD" do not become two rows on the same report.
    def initialize(amount:, currency:)
      code = currency.to_s.strip.upcase

      unless code =~ /\A[A-Z]{3}\z/
        raise InvalidEventError,
              "revenue currency #{currency.inspect} is not a three-letter ISO 4217 code, such as USD or GBP"
      end

      @amount = amount
      @currency = code
    end

    # Renders the wire shape. The key names are the server's and are not
    # configurable, so they live in exactly one place.
    def to_wire
      { "amount" => amount, "currency" => currency }
    end
  end
end
